### PR TITLE
[Feature] Add attention module of CBAM

### PR DESCRIPTION
**Pull request by kitecats**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add attention module of CBAM

## Modification

1. Add attention module of CBAM in mmyolo/models/layers/cbam_layer.py.
2. Add tutorials on the use of the attention module in how_to.

### CODE REVIEW

File: mmyolo/models/layers/cbam_layer.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        avg_out = torch.mean(x, dim=1, keepdim=True)
+        max_out, _ = torch.max(x, dim=1, keepdim=True)
+        out = torch.cat([avg_out, max_out], dim=1)
+        out = self.conv(out)
+        return out
+
+
+@MODELS.register_module()
+class CBAMLayer(BaseModule):
+    """Convolutional Block Attention Module.
```
Comment (Nioolek) [inline mmyolo/models/layers/cbam_layer.py:59]: add arxiv link

**Comment by PeterH0323** [comment]
Pls fix the lint

File: mmyolo/models/layers/cbam_layer.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmyolo.registry import MODELS
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+
+class ChannelAttention(nn.Module):
+    def __init__(self, channels, ratio=16):
```
Comment (hhaAndroid) [inline mmyolo/models/layers/cbam_layer.py:10]: docstring  typehint

Comment (hhaAndroid) [reply mmyolo/models/layers/cbam_layer.py:10]: unit test

File: mmyolo/models/layers/cbam_layer.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmyolo.registry import MODELS
```
Comment (hhaAndroid) [inline mmyolo/models/layers/cbam_layer.py:2]: 新建  mmyolo/models/plugins 文件夹，然后把 cbam 放到里面就行

Comment (hhaAndroid) [inline mmyolo/models/layers/cbam_layer.py:2]: 文件名从 cbam_layer 改成cbam 就行

File: docs/zh_cn/advanced_guides/how_to.md [hunk 4]
```diff
@@ -25,6 +25,22 @@ model = dict(
 
 `cfg` 参数表示插件的具体配置， `stages` 参数表示是否在 backbone 对应的 stage 后面增加插件，长度需要和 backbone 的 stage 数量相同。
 
+### 使用注意力模块
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:28]: 文档的话，在 docs/zh_cn/advanced_guides 里面新建一个 plugins.md ，然后后续所有描述都放里面

Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:28]: how-to 里面可以链接一下路径，告诉用户文档在哪里

File: mmyolo/models/plugins/cbam.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+
+class ChannelAttention(nn.Module):
+    def __init__(self,
+                 channels: int,
+                 ratio: int = 16):
+        super(ChannelAttention, self).__init__()
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:13]:
```suggestion
        super().__init__()
```

File: mmyolo/models/plugins/cbam.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmyolo.registry import MODELS
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+
+class ChannelAttention(nn.Module):
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:9]: docstring  typehint

File: mmyolo/models/plugins/cbam.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        super(ChannelAttention, self).__init__()
+        self.avg_pool = nn.AdaptiveAvgPool2d(1)
+        self.max_pool = nn.AdaptiveMaxPool2d(1)
+        self.conv1 = ConvModule(
+            in_channels=channels,
+            out_channels=int(channels / ratio),
+            kernel_size=1,
+            stride=1,
+            conv_cfg=None,
+            act_cfg=dict(type='ReLU', inplace=True))
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:22]: 这种参数最好设置为可配置

File: mmyolo/models/plugins/cbam.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+from mmengine.model import BaseModule
+
+
+class ChannelAttention(nn.Module):
+    def __init__(self,
+                 channels: int,
+                 ratio: int = 16):
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:12]:
```suggestion
                 reduce_factor: Union[int,float[ = 16):
```

File: mmyolo/models/plugins/cbam.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        super(ChannelAttention, self).__init__()
+        self.avg_pool = nn.AdaptiveAvgPool2d(1)
+        self.max_pool = nn.AdaptiveMaxPool2d(1)
+        self.conv1 = ConvModule(
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:16]: conv1 和 conv2 写成 nn.seq 比较好

File: mmyolo/models/plugins/cbam.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        super(SpatialAttention, self).__init__()
+        assert kernel_size in (3, 7), 'kernel size must be 3 or 7'
+        self.conv = ConvModule(
+            in_channels=2,
+            out_channels=1,
+            kernel_size=kernel_size,
+            stride=1,
+            padding=kernel_size // 2,
+            conv_cfg=None,
+            act_cfg=dict(type='Sigmoid'))
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:50]: same as above

File: mmyolo/models/plugins/cbam.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        avgpool_out = self.conv2(self.conv1(self.avg_pool(x)))
+        maxpool_out = self.conv2(self.conv1(self.max_pool(x)))
+        out = self.sigmoid(avgpool_out + maxpool_out)
+        return out
+
+
+class SpatialAttention(nn.Module):
+    def __init__(self, kernel_size: int = 7):
+        super(SpatialAttention, self).__init__()
+        assert kernel_size in (3, 7), 'kernel size must be 3 or 7'
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:42]: why?

File: mmyolo/models/plugins/cbam.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        avg_out = torch.mean(x, dim=1, keepdim=True)
+        max_out, _ = torch.max(x, dim=1, keepdim=True)
+        out = torch.cat([avg_out, max_out], dim=1)
+        out = self.conv(out)
+        return out
+
+
+@MODELS.register_module()
+class CBAMLayer(BaseModule):
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:61]:
```suggestion
class CBAM(BaseModule):
```

File: mmyolo/models/plugins/cbam.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+          kernel_size (int): The size of the convolution kernel in
+              SpatialAttention. The value can only be 3 or 7. Default: 7.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 ratio: int = 16,
+                 kernel_size: int = 7
+                 ):
+        super(CBAMLayer, self).__init__()
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:77]:
```suggestion
        super().__init__()
```

File: docs/zh_cn/advanced_guides/plugins.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# 更多的插件情况
+
+目前`MMYOLO`支持了如下插件：
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/plugins.md:3]: 中英文要空格

File: docs/zh_cn/advanced_guides/plugins.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+<details open>
+<summary><b>支持的插件</b></summary>
+
+- [x] [CBAM](mmyolo/models/plugins)
+- [x] [GeneralizedAttention](https://github.com/open-mmlab/mmcv/blob/b622fb2e29f44d64a704b91a07b659ef7f6a9397/mmcv/cnn/bricks/generalized_attention.py#L14)
+- [x] [NonLocal2d](https://github.com/open-mmlab/mmcv/blob/b622fb2e29f44d64a704b91a07b659ef7f6a9397/mmcv/cnn/bricks/non_local.py#L219)
+- [x] [ContextBlock](https://github.com/open-mmlab/mmcv/blob/b622fb2e29f44d64a704b91a07b659ef7f6a9397/mmcv/cnn/bricks/context_block.py#L19)
+
+</details>
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/plugins.md:12]: 如何使用可以以 yolov5 为例，给出具体用法

File: mmyolo/models/plugins/cbam.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+        max_out, _ = torch.max(x, dim=1, keepdim=True)
+        out = torch.cat([avg_out, max_out], dim=1)
+        out = self.conv(out)
+        return out
+
+
+@MODELS.register_module()
+class CBAMLayer(BaseModule):
+    """Convolutional Block Attention Module.
+      arxiv link: https://arxiv.org/abs/1807.06521v2
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:63]:
```suggestion
      arxiv link: https://arxiv.org/abs/1807.06521v2
      
```

File: mmyolo/models/plugins/cbam.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+            by this dict. If act_cfg is a sequence of dicts, the first
+            activation layer will be configurated by the first dict and the
+            second activation layer will be configurated by the second dict.
+            Default: (dict(type='ReLU'), dict(type='Sigmoid'))
+    """
+
+    def __init__(self,
+                 channels: int,
+                 ratio: int = 16,
+                 act_cfg: Union[int, Sequence[int]] = (dict(type='ReLU'),
```
Comment (Nioolek) [inline mmyolo/models/plugins/cbam.py:30]: type=relu,inplace=True

File: mmyolo/models/plugins/cbam.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+        self.conv1 = ConvModule(
+            in_channels=channels,
+            out_channels=int(channels / ratio),
+            kernel_size=1,
+            stride=1,
+            conv_cfg=None,
+            act_cfg=None)
+
+        self.activate1 = MODELS.build(act_cfg[0])
```
Comment (Nioolek) [inline mmyolo/models/plugins/cbam.py:49]: Activate可以合并到convmodule中

File: mmyolo/models/plugins/cbam.py (hunk 17 above)
Comment (Nioolek) [reply mmyolo/models/plugins/cbam.py:30]: Union[int, Sequence[int]]？

File: tests/test_models/test_plugins/test_cbam.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+class TestCBAM(TestCase):
+
+    def test_init(self):
+        with pytest.raises(AssertionError):
+            # act_cfg in ChannelAttention sequence length must equal to 2
+            CBAM(
+                in_channels=16,
+                act_cfg=dict(
+                    ChannelAttention=(dict(type='ReLU'), ),
```
Comment (PeterH0323) [inline tests/test_models/test_plugins/test_cbam.py:22]:
```suggestion
                    ChannelAttention=(dict(type='ReLU')),
```

File: docs/zh_cn/advanced_guides/plugins.md [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+        plugins=[
+            dict(
+                cfg=dict(
+                    type='GeneralizedAttention',
+                    spatial_range=-1,
+                    num_heads=8,
+                    attention_type='0011',
+                    kv_stride=2),
+                stages=(False, False, True, True)),
+        ], ))
```
Comment (PeterH0323) [inline docs/zh_cn/advanced_guides/plugins.md:20]:
```suggestion
                stages=(False, False, True, True))
        ]))
```

File: mmyolo/models/plugins/cbam.py [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+            by this dict. If act_cfg is a sequence of dicts, the first
+            activation layer will be configurated by the first dict and the
+            second activation layer will be configurated by the second dict.
+            Default: (dict(type='ReLU'), dict(type='Sigmoid'))
+    """
+
+    def __init__(self,
+                 channels: int,
+                 ratio: int = 16,
+                 act_cfg: Union[dict,
```
Comment (hhaAndroid) [inline mmyolo/models/plugins/cbam.py:32]: act_cfg 里面套两个激活，其实没有这种写法，不推荐

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**